File: android/src/main/java/com/expobraintree/CardPaymentExecutor.kt
package com.expobraintree

import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.fragment.app.FragmentActivity
import com.braintreepayments.api.card.Card
import com.braintreepayments.api.card.CardClient
import com.braintreepayments.api.card.CardResult
import com.braintreepayments.api.threedsecure.ThreeDSecureClient
import com.braintreepayments.api.threedsecure.ThreeDSecureLauncher
import com.braintreepayments.api.threedsecure.ThreeDSecureLauncherCallback
import com.braintreepayments.api.threedsecure.ThreeDSecurePaymentAuthRequest
import com.braintreepayments.api.threedsecure.ThreeDSecurePaymentAuthResult
import com.braintreepayments.api.threedsecure.ThreeDSecureRequest
import com.braintreepayments.api.threedsecure.ThreeDSecureResult

class CardPaymentExecutor(
    private val context: Context,
    listener: PaymentExecutorListener,
    args: BasePaymentArgs
) : BasePaymentExecutor(args, listener), ThreeDSecureLauncherCallback {

    private val cardClient: CardClient
    private var threeDSecureClient: ThreeDSecureClient? = null
    private var threeDSecureLauncher: ThreeDSecureLauncher? = null
    private var currentCardNonce: String? = null

    init {
        Log.d(TAG, "[init] Creating CardClient, tokenPresent=${args.clientToken.isNotEmpty()}")
        cardClient = CardClient(
            context = context,
            authorization = args.clientToken
        )

        // Initialize 3DS client and launcher early if 3DS is enabled
        val paymentMethod = args.paymentMethod as? PaymentMethod.Card
        if (paymentMethod?.use3DSecure == true && context is FragmentActivity) {
            Log.d(TAG, "[init] Initializing 3DS client and launcher early")
            threeDSecureClient = ThreeDSecureClient(
                context = context,
                authorization = args.clientToken
            )
            threeDSecureLauncher = ThreeDSecureLauncher(context, this)
        }
    }

    override fun requestPayment(activity: FragmentActivity) {
        val paymentMethod = args.paymentMethod as? PaymentMethod.Card
        if (paymentMethod == null) {
            Log.e(TAG, "[requestPayment] Invalid payment method")
            handleError("Invalid payment method", "Expected Card")
            return
        }

        Log.d(TAG, "[requestPayment] Starting card tokenization, use3DSecure=${paymentMethod.use3DSecure}")

        val card = Card().apply {
            number = paymentMethod.cardNumber
            expirationMonth = paymentMethod.expirationMonth
            expirationYear = paymentMethod.expirationYear
            if (!paymentMethod.cvv.isNullOrEmpty()) {
                cvv = paymentMethod.cvv
            }
            if (!paymentMethod.postalCode.isNullOrEmpty()) {
                postalCode = paymentMethod.postalCode
            }
        }

        Log.d(TAG, "[requestPayment] Card prepared, tokenizing...")

        cardClient.tokenize(card) { cardResult ->
            Log.d(TAG, "[requestPayment] tokenize callback, result type: ${cardResult::class.simpleName}")

            when (cardResult) {
                is CardResult.Success -> {
                    val nonce = cardResult.nonce
                    Log.d(TAG, "[requestPayment] Card tokenization success, cardType=${nonce.cardType}")

                    if (paymentMethod.use3DSecure) {
                        Log.d(TAG, "[requestPayment] 3DS enabled, starting verification")
                        currentCardNonce = nonce.string
                        requestThreeDSecure(activity, nonce)
                    } else {
                        Log.d(TAG, "[requestPayment] 3DS not required, returning nonce")
                        handleSuccessWithCardMethod(nonce.string)
                    }
                }
                is CardResult.Failure -> {
                    Log.e(TAG, "[requestPayment] Card tokenization failure: ${cardResult.error.message}")
                    handleError(cardResult.error.message, cardResult.error.localizedMessage)
                }
            }
        }
    }

    private fun requestThreeDSecure(activity: FragmentActivity, cardNonce: com.braintreepayments.api.card.CardNonce) {
        val paymentMethod = args.paymentMethod as PaymentMethod.Card
        Log.d(TAG, "[requestThreeDSecure] Starting 3DS verification")

        // Ensure 3DS client and launcher are initialized
        if (threeDSecureClient == null) {
            Log.d(TAG, "[requestThreeDSecure] Late initialization of 3DS client")
            threeDSecureClient = ThreeDSecureClient(
                context = context,
                authorization = args.clientToken
            )
        }

        if (threeDSecureLauncher == null) {
            Log.e(TAG, "[requestThreeDSecure] ThreeDSecureLauncher not initialized - this should not happen!")
            handleError("3D Secure initialization error", "Launcher not initialized early enough")
            return
        }

        Log.d(TAG, "[requestThreeDSecure] cardType: ${cardNonce.cardType}, amount: ${paymentMethod.amount}")

        val threeDSecureRequest = ThreeDSecureRequest(
            nonce = cardNonce.string,
            amount = paymentMethod.amount,
            email = args.email
        )

        Log.d(TAG, "[requestThreeDSecure] Creating 3DS payment auth request")

        threeDSecureClient!!.createPaymentAuthRequest(
            context = activity,
            request = threeDSecureRequest
        ) { paymentAuthRequest ->
            Log.d(TAG, "[requestThreeDSecure] createPaymentAuthRequest callback, type: ${paymentAuthRequest::class.simpleName}")

            when (paymentAuthRequest) {
                is ThreeDSecurePaymentAuthRequest.Failure -> {
                    Log.e(TAG, "[requestThreeDSecure] 3DS auth request failure: ${paymentAuthRequest.error.message}")
                    handleError(paymentAuthRequest.error.message, paymentAuthRequest.error.localizedMessage)
                }
                is ThreeDSecurePaymentAuthRequest.LaunchNotRequired -> {
                    Log.d(TAG, "[requestThreeDSecure] 3DS launch not required, using nonce directly")
                    handleSuccessWithCardMethod(paymentAuthRequest.nonce.string)
                }
                is ThreeDSecurePaymentAuthRequest.ReadyToLaunch -> {
                    Log.d(TAG, "[requestThreeDSecure] 3DS ready to launch, launching...")
                    threeDSecureLauncher?.launch(paymentAuthRequest)
                }
            }
        }
    }

    override fun onThreeDSecurePaymentAuthResult(paymentAuthResult: ThreeDSecurePaymentAuthResult) {
        Log.d(TAG, "[onThreeDSecurePaymentAuthResult] Received 3DS auth result, tokenizing...")

        threeDSecureClient?.tokenize(paymentAuthResult) { threeDSecureResult ->
            Log.d(TAG, "[onThreeDSecurePaymentAuthResult] tokenize callback, type: ${threeDSecureResult::class.simpleName}")

            when (threeDSecureResult) {
                is ThreeDSecureResult.Success -> {
                    val threeDSecureNonce = threeDSecureResult.nonce
                    val threeDSecureInfo = threeDSecureNonce.threeDSecureInfo

                    Log.d(TAG, "[onThreeDSecurePaymentAuthResult] 3DS success, liabilityShiftPossible=${threeDSecureInfo.liabilityShiftPossible}, liabilityShifted=${threeDSecureInfo.liabilityShifted}")

                    if (!threeDSecureInfo.liabilityShiftPossible) {
                        Log.e(TAG, "[onThreeDSecurePaymentAuthResult] Liability shift not possible")
                        handleError("3D Secure liability shift not possible", "3D Secure liability shift not possible")
                        return@tokenize
                    }

                    if (!threeDSecureInfo.liabilityShifted) {
                        Log.e(TAG, "[onThreeDSecurePaymentAuthResult] Liability not shifted")
                        handleError("3D Secure liability not shifted", "3D Secure liability not shifted")
                        return@tokenize
                    }

                    Log.d(TAG, "[onThreeDSecurePaymentAuthResult] 3DS verification successful")
                    handleSuccessWithCardMethod(threeDSecureNonce.string)
                }
                is ThreeDSecureResult.Failure -> {
                    Log.e(TAG, "[onThreeDSecurePaymentAuthResult] 3DS failure: ${threeDSecureResult.error.message}")
                    handleError(threeDSecureResult.error.message, threeDSecureResult.error.localizedMessage)
                }
                is ThreeDSecureResult.Cancel -> {
                    Log.d(TAG, "[onThreeDSecurePaymentAuthResult] 3DS cancelled by user")
                    handleCancel()
                }
            }
        }
    }

    private fun handleSuccessWithCardMethod(nonce: String) {
        val paymentMethod = args.paymentMethod as PaymentMethod.Card
        handleSuccess(
            nonce = nonce,
            amount = paymentMethod.amount,
            currency = if (paymentMethod.currency.isNotEmpty()) paymentMethod.currency else "USD",
            paymentType = "Card"
        )
    }

    override fun onResume(intent: Intent) {
        Log.d(TAG, "[onResume] Activity resumed")
    }

    override fun onDestroy() {
        Log.d(TAG, "[onDestroy] Cleaning up")
        threeDSecureLauncher = null
        threeDSecureClient = null
    }
}

File: ios/CardPaymentExecutor.swift
//
//  CardPaymentExecutor.swift
//  expo-braintree
//
//  Card payment executor with 3DS support
//

import Braintree
import Foundation
import UIKit

class CardPaymentExecutor: NSObject, BasePaymentExecutor, BTThreeDSecureRequestDelegate {
  let args: BasePaymentArgs
  weak var listener: PaymentExecutorListener?
  private var cardClient: BTCardClient?
  private var threeDSecureClient: BTThreeDSecureClient?
  private var currentCardNonce: String?
  private weak var viewController: UIViewController?

  init(args: BasePaymentArgs, listener: PaymentExecutorListener?) {
    self.args = args
    self.listener = listener
    super.init()
  }

  func requestPayment(viewController: UIViewController) {
    self.viewController = viewController

    guard case .card(
      let cardNumber,
      let expirationMonth,
      let expirationYear,
      let cvv,
      let postalCode,
      let use3DSecure,
      _,
      _,
      _
    ) = args.paymentMethod else {
      handleError(message: "Invalid payment method", localizedMessage: "Expected Card")
      return
    }

    guard let apiClient = BTAPIClient(authorization: args.clientToken) else {
      handleError(
        message: ERROR_TYPES.API_CLIENT_INITIALIZATION_ERROR.rawValue,
        localizedMessage: "Failed to initialize Braintree API client"
      )
      return
    }

    cardClient = BTCardClient(apiClient: apiClient)

    if use3DSecure {
      threeDSecureClient = BTThreeDSecureClient(apiClient: apiClient)
    }

    let card = BTCard()
    card.number = cardNumber
    card.expirationMonth = expirationMonth
    card.expirationYear = expirationYear
    if let cvvValue = cvv, !cvvValue.isEmpty {
      card.cvv = cvvValue
    }
    if let postalCodeValue = postalCode, !postalCodeValue.isEmpty {
      card.postalCode = postalCodeValue
    }
    card.shouldValidate = false

    cardClient?.tokenize(card) { [weak self] (cardNonce, error) in
      guard let self = self else { return }

      if let cardNonce = cardNonce {
        if use3DSecure {
          self.currentCardNonce = cardNonce.nonce
          self.requestThreeDSecure(cardNonce: cardNonce)
        } else {
          self.handleSuccessWithCardNonce(cardNonce: cardNonce)
        }
      } else if let error = error {
        self.handleError(
          message: ERROR_TYPES.CARD_TOKENIZATION_ERROR.rawValue,
          localizedMessage: error.localizedDescription
        )
      }
    }
  }

  private func requestThreeDSecure(cardNonce: BTCardNonce) {
    guard case .card(_, _, _, _, _, _, let amount, _, let requestorAppURL) = args.paymentMethod else {
      handleError(message: "Invalid payment method", localizedMessage: "Expected Card")
      return
    }

    guard let secureClient = threeDSecureClient else {
      handleError(
        message: ERROR_TYPES.API_CLIENT_INITIALIZATION_ERROR.rawValue,
        localizedMessage: "3D Secure client not initialized"
      )
      return
    }

    let threeDSecureRequest = BTThreeDSecureRequest()
    let formatter = NumberFormatter()
    formatter.generatesDecimalNumbers = true
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.numberStyle = .decimal

    threeDSecureRequest.amount = formatter.number(from: amount) as? NSDecimalNumber ?? 0
    threeDSecureRequest.nonce = cardNonce.nonce
    threeDSecureRequest.email = args.email
    threeDSecureRequest.threeDSecureRequestDelegate = self
    threeDSecureRequest.requestorAppURL = requestorAppURL

    secureClient.startPaymentFlow(threeDSecureRequest) { [weak self] (threeDSecureResult, error) in
      guard let self = self else { return }

      if let threeDSecureResult = threeDSecureResult,
         let tokenizedCard = threeDSecureResult.tokenizedCard {

        let threeDSecureInfo = tokenizedCard.threeDSecureInfo

        if !threeDSecureInfo.liabilityShiftPossible {
          self.handleError(
            message: ERROR_TYPES.THREE_D_SECURE_NOT_ABLE_TO_SHIFT_LIABILITY.rawValue,
            localizedMessage: "3D Secure liability shift not possible"
          )
          return
        }

        if !threeDSecureInfo.liabilityShifted {
          self.handleError(
            message: ERROR_TYPES.THREE_D_SECURE_LIABILITY_NOT_SHIFTED.rawValue,
            localizedMessage: "3D Secure liability not shifted"
          )
          return
        }

        self.handleSuccessWithCardNonce(cardNonce: tokenizedCard)
      } else if let error = error {
        if let threeDSecureError = error as? BTThreeDSecureError, threeDSecureError == .canceled {
          self.handleCancel()
        } else if (error as NSError).code == BTThreeDSecureError.canceled.errorCode {
          self.handleCancel()
        } else {
          self.handleError(
            message: ERROR_TYPES.THREE_D_SECURE_AUTHENTICATION_FAILED.rawValue,
            localizedMessage: error.localizedDescription
          )
        }
      } else {
        self.handleError(
          message: ERROR_TYPES.THREE_D_SECURE_VERIFICATION_FAILED.rawValue,
          localizedMessage: "3D Secure verification failed"
        )
      }
    }
  }

  private func handleSuccessWithCardNonce(cardNonce: BTCardNonce) {
    guard case .card(_, _, _, _, _, _, let amount, let currency, _) = args.paymentMethod else {
      handleError(message: "Invalid payment method", localizedMessage: "Expected Card")
      return
    }

    let additionalData = prepareBTCardNonceResult(cardNonce: cardNonce) as? [String: Any]

    handleSuccess(
      nonce: cardNonce.nonce,
      amount: amount,
      currency: currency.isEmpty ? "USD" : currency,
      paymentType: "Card",
      additionalData: additionalData
    )
  }

  func onLookupComplete(_ request: BTThreeDSecureRequest, lookupResult: BTThreeDSecureResult, next: @escaping () -> Void) {
    next()
  }

  func onResume() {
  }

  func onDestroy() {
    cardClient = nil
    threeDSecureClient = nil
    currentCardNonce = nil
  }
}
